### DISCUSSION TITLE
Home assistant

### DISCUSSION THREAD

**Discussion by matthew73210** [mmcc-xx/BirdCAGE#3] [Ideas]
Now that you've managed the unfathomable, a step forward to getting a bird analysing add-on for HA. Cheers

**Comment by mmcc-xx** [comment]
What did you have in mind? It would be pretty straightforward to send out MQTT messages with detections - would that be a good first step?

**Reply by DemBones78** [reply]
i dont know , what is the benefit of sending all the detections?  my PI gets 4k+ detections aday with ! mic.  i have wanted this to pul in 8 cams over 40 acres . when that happens 50k detection aday would probably kill HA..    how about monitoring , like a watch dog. kinda how the phones show up/.  IDK maybe it not a big deal if HA does not log most.  https://app.birdweather.com/stations/644

**Reply by matthew73210** [reply]
You don't have to send every detection. Saying that HA is more than capable, I have loads of sensors with second reports. The easiest option would be to do an hourly report (or based on what reporting you want) and send it over mqtt. The idea being that it would open up the whole sphere to HA as an add on.
If my memory serves me well there's a certain method if adding addons so that they appear on HA. Or you can do a custom store. I'll post on the HA forums and see what they say, I'll point back to your repo.

**Reply by mmcc-xx** [reply]
Please also point me at the discussion you start on the HA forums - I’d be interested to see the discussion. I have another repo here that does works with frigate and does species ID of snap shots from Frigate, so it might be interesting to integrate that data together somehow.

**Reply by mmcc-xx** [reply]
Meanwhile I'll post something on the HA subreddit to get some thoughts.

**Reply by matthew73210** [reply]
> Please also point me at the discussion you start on the HA forums - I’d be interested to see the discussion. I have another repo here that does works with frigate and does species ID of snap shots from Frigate, so it might be interesting to integrate that data together somehow.

Here's the original post I created a while back : https://community.home-assistant.io/t/birdnet-pi/444654/6
Could you add the link to the subreddit on the main page please?

**Reply by brucewinter** [reply]
Sending detections out to mqtt is a terrific idea!!   

Would love to integrate it with home assistant, to enable announcements and replays of unique/interesting bird detections.  Could also enable plots of detections that integrate nicely with on the ha dashboards.  

I think all we would need in the mqtt message would be the same data posted to the "Recent Detection" page:  time, stream name, common name, confidence, and a buffer of the audio for playback.

**Comment by brucewinter** [comment]
Thanks for adding the new mqtt option!!   Working great so far, but I do have one issue/question.

I'm not sure how to save/use the mp3 data.  I've tried saving it multiple ways (e.g. node-red file writes, manual emacs copy/write), but the resulting file does not play, either as a .mp3 or .wav file.  Pretty sure I need to re-encode the data somehow.  When I look at playable .mp3/.wav files, I see binary data in them (e.g. \###  characters), but when I look at the mqtt data, it is all ascii.   Any suggestions?

**Reply by mmcc-xx** [reply]
It is base64 encoded - the same way, I think, that Frigate sends images in its snapshot MQTT publications. 

Does the info below help? If not I'll dig into it more. I haven't tried decoding any of the published binaries, so I certainly could have messed it up.

"There is a base64 Node-RED node ([node-red-node-base64](https://flows.nodered.org/node/node-red-node-base64)) that will convert a base64 encoded string into a NodeJS binary buffer.

If you run your string through this before sending it to a file node you will get a correctly formatted image."
https://stackoverflow.com/questions/46074346/howto-save-a-base64-encoded-image-using-node-red

**Comment by brucewinter** [comment]
Saving the file with base64 encoding did the trick!    I'm having fun "re-tweeting" bird calls via inside house speakers  :)   

Thanks for adding the mqtt option and thanks for the tip.

**Reply by mmcc-xx** [reply]
This is retweeting I can get behind

**Reply by matthew73210** [reply]
Having fun with nocturnal birds waking you up?

**Reply by brucewinter** [reply]
Funny you should mention that!   I got woken up this morning at 5 am by some now-indoor bard owl hoots.   Will be putting in a 'is sleeping?' test in my node-red code later today ;)

**Comment by brucewinter** [comment]
Love the new Weekly/Monthly/Annual reports!   

One other report idea, detection report by stream.   I've got 6 streams happily running now, from cameras I have in different places in my backyard forest.  It would be fun to see where the birds hang out the most.

I would guess most people would only have 1 or 2 streams, so it might not be worth the effort, feel free to pass on this idea.

**Reply by mmcc-xx** [reply]
Duly noted! My next tasks are finally implementing a way to apply settings changes without restarting the app, and doing some documentation. But setting up filtering by stream makes total sense.

**Comment by Nixon506E** [comment]
I would like a setting in the MQTT section to add 'retain' to the messages. This would allow me to reload the last detected bird when restarting HASS.

Also hoping that the MQTT messages could include the occurrence score for the species as well so that I can set up automations in HASS rather than relying on the detection filters.

**Reply by mmcc-xx** [reply]
That's for the suggestion. That makes sense.

**Reply by mmcc-xx** [reply]
MQTT pubs now have the retain flag set, and the payload includes the occurrence score.

**Comment by matthew73210** [comment]
Comming back to this thread, this is what I was thinking originally about https://developers.home-assistant.io/docs/add-ons/tutorial/. Now that you have apprise and mqtt, maybe no need.

**Reply by matthew73210** [reply]
I tried but couldn't get it running as it wants to access forbidden folders on HA (tmp etc...), can these not be virtual ones inside the birdcage dir ?

**Reply by matthew73210** [reply]
Progress, so with minimal changes (only to the ips). I can get the thing to install through homeassistant add-ons, now I'm having issues with the backend container, it starts but hangs (as far as I can tell). Is it waiting for something in particular?

**Reply by mmcc-xx** [reply]
I’ll have to look at how ha addons work. The backend needs redis and needs the birdnet server container. I’d guess it isn’t talking to redis.

**Reply by mmcc-xx** [reply]
Are you able to get at the logs from birdcage_backend?

**Reply by matthew73210** [reply]
I'm now able to get logs, basicly I'm going through another container. It's complaining about the mount source being non writable. But the filesystem is mounted as RW...

**Reply by mmcc-xx** [reply]
I wonder if it would make any different to mount the tmp directory as a normal directory instead of a temp filesystem?
So instead of..
```
    tmpfs:
      - /tmp:size=16M
```

Do...
```
    volumes:
      - "./tmp:/tmp"
      - "./detections:/detections"
      - "./db:/db"
      - "/etc/localtime:/etc/localtime:ro"
      - "/etc/timezone:/etc/timezone:ro"
```

**Reply by matthew73210** [reply]
Manged to get to the first page :  modified the following in docker-compose
`    volumes_from:
      - container:addon_local_birdcage_bird
    volumes:
      - "/etc/localtime:/etc/localtime:ro"
      - "/etc/timezone:/etc/timezone:ro"
    depends_on:
      - redis
    networks:
      - birdcage_net`
 And mounted share as rw from the config.yaml. Created two dirs as db and detections.
Are they expecting to be in birdcage or does it not really matter ?
Basicly the whole issue is that you have to create a first container which mounts volumes, then run docker-compose which can't mount volumes as rw but can mount from other containers.

**Reply by matthew73210** [reply]
Now i am getting the folowwing :  

birdcage-birdcage_backend-1  | Starting API Server
birdcage-birdcage_backend-1  | Starting celery_worker
birdcage-birdcage_backend-1  | Sleeping indefinitely
birdcage-birdcage_backend-1  | [2023-06-10 20:55:46 +0000] [7] [INFO] Starting gunicorn 20.1.0
birdcage-birdcage_backend-1  | [2023-06-10 20:55:46 +0000] [7] [INFO] Listening at: http://0.0.0.0:7007 (7)
birdcage-birdcage_backend-1  | [2023-06-10 20:55:46 +0000] [7] [INFO] Using worker: sync
birdcage-birdcage_backend-1  | [2023-06-10 20:55:46 +0000] [18] [INFO] Booting worker with pid: 18
birdcage-birdcage_backend-1  | Starting celery worker in celery_worker.py
birdcage-birdcage_backend-1  | Traceback (most recent call last):
birdcage-birdcage_backend-1  |   File "//celery_worker.py", line 5, in <module>
birdcage-birdcage_backend-1  |     app = create_app(init_celery=True)
birdcage-birdcage_backend-1  |   File "/app/__init__.py", line 31, in create_app
birdcage-birdcage_backend-1  |     create_streams_table()
birdcage-birdcage_backend-1  |   File "/app/models/streams.py", line 6, in create_streams_table
birdcage-birdcage_backend-1  |     connection = sqlite3.connect(DATABASE_FILE)
birdcage-birdcage_backend-1  | sqlite3.OperationalError: unable to open database file
birdcage-birdcage_backend-1  | [2023-06-10 20:55:48 +0000] [18] [ERROR] Exception in worker process
birdcage-birdcage_backend-1  | Traceback (most recent call last):
birdcage-birdcage_backend-1  |   File "/usr/local/lib/python3.10/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
birdcage-birdcage_backend-1  |     worker.init_process()
birdcage-birdcage_backend-1  |   File "/usr/local/lib/python3.10/site-packages/gunicorn/workers/base.py", line 134, in init_process
birdcage-birdcage_backend-1  |     self.load_wsgi()
birdcage-birdcage_backend-1  |   File "/usr/local/lib/python3.10/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
birdcage-birdcage_backend-1  |     self.wsgi = self.app.wsgi()
birdcage-birdcage_backend-1  |   File "/usr/local/lib/python3.10/site-packages/gunicorn/app/base.py", line 67, in wsgi
birdcage-birdcage_backend-1  |     self.callable = self.load()
birdcage-birdcage_backend-1  |   File "/usr/local/lib/python3.10/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
birdcage-birdcage_backend-1  |     return self.load_wsgiapp()
birdcage-birdcage_backend-1  |   File "/usr/local/lib/python3.10/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
birdcage-birdcage_backend-1  |     return util.import_app(self.app_uri)
birdcage-birdcage_backend-1  |   File "/usr/local/lib/python3.10/site-packages/gunicorn/util.py", line 359, in import_app
birdcage-birdcage_backend-1  |     mod = importlib.import_module(module)
birdcage-birdcage_backend-1  |   File "/usr/local/lib/python3.10/importlib/__init__.py", line 126, in import_module
birdcage-birdcage_backend-1  |     return _bootstrap._gcd_import(name[level:], package, level)
birdcage-birdcage_backend-1  |   File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
birdcage-birdcage_backend-1  |   File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
birdcage-birdcage_backend-1  |   File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
birdcage-birdcage_backend-1  |   File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
birdcage-birdcage_backend-1  |   File "<frozen importlib._bootstrap_external>", line 883, in exec_module
birdcage-birdcage_backend-1  |   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
birdcage-birdcage_backend-1  |   File "/run.py", line 10, in <module>
birdcage-birdcage_backend-1  |     app = create_app(init_celery=False)
birdcage-birdcage_backend-1  |   File "/app/__init__.py", line 31, in create_app
birdcage-birdcage_backend-1  |     create_streams_table()
birdcage-birdcage_backend-1  |   File "/app/models/streams.py", line 6, in create_streams_table
birdcage-birdcage_backend-1  |     connection = sqlite3.connect(DATABASE_FILE)
birdcage-birdcage_backend-1  | sqlite3.OperationalError: unable to open database file
birdcage-birdcage_backend-1  | [2023-06-10 20:55:48 +0000] [18] [INFO] Worker exiting (pid: 18)
birdcage-birdcage_backend-1  | [2023-06-10 20:55:48 +0000] [7] [INFO] Shutting down: Master
birdcage-birdcage_backend-1  | [2023-06-10 20:55:48 +0000] [7] [INFO] Reason: Worker failed to boot.

**Reply by matthew73210** [reply]
In share I created birdcage/db and detections. updated the following 
      DATABASE_FILE: share/birdcage/db/birdcage.db
      DETECTION_DIR_NAME: share/birdcage/detections
And voila, it works. Lets see if I can get detections

**Reply by matthew73210** [reply]
Okay I can confirm that it works, currently pointing at the youtube page you are using with lat/log set accordingly. Now I need to find the right settings to get the add-on to remove / stop the docker-compose created containers when you uninstall/stop

**Reply by mmcc-xx** [reply]
Whoa that's awesome! Can you post a screenshot? I am completely ignorant of how HA addons work and look. I am stuck in the HA dark ages.

**Reply by matthew73210** [reply]
<img width="1808" alt="Screenshot 2023-06-10 at 11 53 43 PM" src="https://github.com/mmcc-xx/BirdCAGE/assets/19491784/9f8fc824-b868-428b-b327-e8b6cc11ffd3">
The ingress part still needs working on as it point to the wrong adresse, but the manual IP + port works fine

**Reply by matthew73210** [reply]
<img width="1232" alt="Screenshot 2023-06-10 at 11 56 08 PM" src="https://github.com/mmcc-xx/BirdCAGE/assets/19491784/9876b9f3-aa2c-4d4f-8f8a-2d9ed7067325">
<img width="1232" alt="Screenshot 2023-06-10 at 11 56 20 PM" src="https://github.com/mmcc-xx/BirdCAGE/assets/19491784/378fcade-c600-4c76-8237-f119bd916ad7">

**Reply by mmcc-xx** [reply]
If you want to do a pull request I'll happily accept it. Thanks for your work!

**Reply by matthew73210** [reply]
I'm not at home this week, but I'm already thinking of another plan of attack. Basicly going to use the base HA docker image which includes an S6 overlay. Like a supervisor I can run docker-compose up and dow etc... As it uses the docker_api the score of it is 1 out of 7 but that's how its gonna be. Do you think it would work if you changed the IP to ${HOST_IP_ADDRESS} to make it dynamic and not need to touch the docker-compose ?

**Comment by matthew73210** [comment]
Hey,

Okay here's a zip file which contains all you need to get it running as an add-on. You need to create a samba connecting and go to the addons folder, then unzip. Modify parts of the file that protain to ips etc (changed to a random one) and voila.
Because it uses the docker-api it gets the lowest score.
[birdcagev2.zip](https://github.com/mmcc-xx/BirdCAGE/files/11781747/birdcagev2.zip)

**Reply by phillipredshirt** [reply]
Can you explain in more detail how to get this add-on running in HA? I edited the IP addresses in the docker-compose.yml file, changing the IP addresses to the address of my HA machine, put the folder in the addons directory, rebooted Home Assistant and I see nothing in HA.

This is running HAOS on a Pi5

Thanks.
